### DISCUSSION TITLE
react-number-format integration

### DISCUSSION THREAD

**Discussion by keemor** [react-hook-form/react-hook-form#1957] [Q&A] [answered]
Hi,

I'm looking for a way to return floatValue from react-number-format 

I found [this thread](https://github.com/react-hook-form/react-hook-form/issues/572) but I'm not sure if using RHFInput is still valid.

CSB [https://codesandbox.io/s/react-hook-form-number-format-custom-integration-2i82q](https://codesandbox.io/s/react-hook-form-number-format-custom-integration-2i82q?file=/src/index.js:187-206)

```
{
  formattedValue: '$23,234,235.56', //value after applying formatting
  value: '23234235.56', //non formatted value as numeric string 23234235.56, if you are setting this value to state make sure to pass isNumericString prop to true
  floatValue: 23234235.56 //floating point representation. For big numbers it can have exponential syntax
}
```

Right now if you type the number you will get it formattedValue.
![image](https://user-images.githubusercontent.com/1793635/85681662-cc465880-b6cb-11ea-85c1-d83029ea29ae.png)

I would like to avoid formatting data onSubmit and simply return proper value from the component.

Thank you for any suggestions and great job 😁

**Comment by bluebill1049** [comment, marked as answer]
hey @keemor we moved over to `Controller` now, hmmm I think you will have to map the data at submit or you will have to map with defaultValue and change data during each onChange.

**Reply by keemor** [reply]
Changing data during each onChange will break input value formatting, so I think mapping data at submit is the only go.

**Reply by bluebill1049** [reply]
what about this?

```
       <InputStyled
            type="text"
            name={name}
            onBlur={e => {
              const value = e.target.value;
              if (value.length > 0) {
                const valueInNumber = convertStringToNumber(value);
                e.target.value = `$${(Number.isNaN(valueInNumber) ? 0 : valueInNumber).toLocaleString('en-EN')}`;
              }
              onBlur(e);
            }}
            onChange={e => {
              const value = e.target.value;
              const valueInNumber = convertStringToNumber(value);
              if (valueInNumber > 1) {
                e.target.value = `$${(Number.isNaN(valueInNumber) ? 0 : valueInNumber).toLocaleString('en-EN')}`;
              }
              onChange(e);
            }}
            placeholder="$"
            ref={register({
              required,
              max,
              min,
              maxLength,
              minLength,
              errors,
              pattern,
            })}
          />
```

**Reply by keemor** [reply]
@bluebill1049 I'm a little bit lost how it should go along with my example integrating Controller, NumberFormat & TextField 🤔

```
<Controller
  as={
    <NumberFormat
    customInput={TextField}
...
```

**Reply by bluebill1049** [reply]
the example above doesn't even need `Controller`,  it's basically a native input (styled-component).

**Reply by bluebill1049** [reply]
actually never mind, it still has the issue with convert back to a normal number... forget about what i suggested above.

**Reply by keemor** [reply]
That might be also a solution, but I'm already mapping data onSubmit and have nice material look & feel, 

Thank you for your quick responses! 🎉

**Reply by ljosberinn** [reply]
@keemor hey, wanted to let you know we found a solution for that as we also didn't want to reformat in handleSubmit: https://codesandbox.io/s/affectionate-hill-275i3?file=/src/CurrencyInput.tsx

**Reply by farmer911** [reply]
> @keemor hey, wanted to let you know we found a solution for that as we also didn't want to reformat in handleSubmit: https://codesandbox.io/s/affectionate-hill-275i3?file=/src/CurrencyInput.tsx

that so cool, but how can I add rule for the min or max value? I'm don't know how to do that

**Comment by HadiSDev** [comment]
I might have a more clean solution:
```
        <Controller name="test" control={control} render={(props) => {
          return <NumberFormat value={props.value} onValueChange={(target)=>{
            props.onChange();
            setValue("test", target.floatValue);
          }} onBlur={props.onBlur}/>
        }}/>
```

**Reply by bluebill1049** [reply]
perfect!
